src/accountsdb/accounts_file.zig

```diff
@@ -253,7 +252,7 @@ pub const AccountFile = struct {
             null,
             file_size,
             std.posix.PROT.READ | std.posix.PROT.WRITE,
-            std.posix.MAP{ .TYPE = .SHARED },
+            std.posix.MAP{ .TYPE = .PRIVATE },
```

Comment:
once an account file is created its never modified - so its better to keep it private imo

Reply:
To clarify what the difference between `SHARED` and `PRIVATE` is.
- `PRIVATE` makes it impossible for other processes to see into the page and it prevents mutations to the page from being reflected in the underlying file.
- `SHARED` does allow the above two things.

There are other differences, but this is what's important to us.
There is *no* difference in performance between PRIVATE and SHARED.

The question we need to ask ourselves is, do we need the underlying account file to see the mutations? Do they need to persist across runs of the validator?

Reply:
> once an account file is created its never modified - so its better to keep it private imo

I forgot there are two mmaps used in this process. The file is created with one mmap, where it is written to, and then another mmap is used immediately after. In that case, it seems fine to use SHARED for the first mmap and then PRIVATE for the second mmap.

Reply:
> Yes there is. Did you read my comment? I already explained why PRIVATE uses more memory or storage when the data is modified. You're directly contradicting me without addressing my points or providing any kind of argument supporting your claim.

My bad, I should have addressed the arguments more clearly.

> If the system needs to free up memory by moving the modified data from ram to disk, with SHARED it just writes to the original file. But with PRIVATE, it needs to swap the memory to the swap partition or the swap file.

I do not understand the argument here, what is the performance difference between flushing to the original file and flushing to a swap file/partition?

>  So now you have the account on disk in two separate places. 

No you do not, swapping happens per-page not per-mapping. See [here](https://www.kernel.org/doc/gorman/html/understand/understand014.html), specifically this part:
> Strictly speaking, Linux does not swap as “swapping” refers to coping an entire process address space to disk and “paging” to copying out individual pages. Linux actually implements paging as modern hardware supports it, but traditionally has called it swapping in discussions and documentation. 

> This means increased storage usage. You might also hit OOM if the swap is not sufficiently large to store the modified account data.

This could be true, I cannot argue it. I think you're correct that using `PRIVATE` could lead to a bit more memory usage, however it's important to understand that read-only pages, or more specifically pages that have not been written to, are clean in nature and can be dropped by the kernel when needed.

From what I understand, our use case has us writing to a large portion of the file, here
https://github.com/Syndica/sig/blob/8f4ddae3a0297377224a6e6460b6b9b959271159/src/accountsdb/db.zig#L1472-L1476
and then never writing to it again, only reading.

I wonder if this means those dirty pages won't be able to get dropped and will remain in memory. Unsure, would need to do some exact testing for that.

Reply:
I'm satisfied with using PRIVATE here since we don't need to make changes to the data. But just to clarify what I meant...

> what is the performance difference between flushing to the original file and flushing to a swap file/partition?

The difference is storage usage. Flushing to the file doesn't use any extra space. It recycles the space already used by the file. Flushing to swap means you need to consume some swap space, in addition to the space the file is already using.

> > So now you have the account on disk in two separate places.
> 
> No you do not, swapping happens per-page not per-mapping.

I was not trying to say that the entire file must be swapped all at once. My point was just that there are now two locations on the disk that are responsible for storing data from the file or the mapping, rather than solely using the file for that purpose.